### ISSUE TITLE
Vivado shows critical warning [Synth 8-295] found timing loop

### ISSUE THREAD

**Issue by Edgar-Conzen** [VectorBlox/orca#4]
Hi.  Not sure if this is a problem with your source code or my tooling / level of experience.
I am using Vivado 2017.1 and your sources from the zip file "orca-master.zip\orca-master\rtl\*.*"

Everything synthesizes fine except for the following critical warnings
(Messages edited manually for your convenience)

[Synth 8-295] found timing loop. ["../utils.vhd":113]
    return pc & target & taken & flush & enable;

[Synth 8-295] found timing loop. ["../orca_core.vhd":136]
  d_valid <= if_valid_out and not pipeline_flush;

[Synth 8-295] found timing loop. ["../execute.vhd":14]
   entity execute is

Is this a known error or even intended? Did you try Xilinx Vivado? I only see Altera and Lattice as FAMILY.

Thank you very much.

**Comment by vanjoe** [comment]
Hi,

Sorry we took so long to get back to you, we were in the middle of updating the repository, we have now done so. Please check if your issue still exists.

Thanks

**Comment by Edgar-Conzen** [comment]
Now I have tried your recent code and the previous one with Vivado 2016.3 / 2017.1 and 2017.2 and the problem remains. 

As soon as I add a clock constraint like 
> create_clock -add -name clk -period 10.00 -waveform {0 5} [get_nets { clk }];

critical warnings emerge. With your new code this are:

>Synthesis[Synth 8-295] found timing loop. ["../utils.vhd":127]
>[Synth 8-295] found timing loop. ["../utils.vhd":127]
>[Synth 8-295] found timing loop. ["../decode.vhd":85]
>[Synth 8-295] found timing loop. ["..orca.vhd":8]
>[Synth 8-295] found timing loop. ["../orca.vhd":8]

Is there something wrong with my constraint?

Thanks!

**Comment by rdeiaco** [comment]
Hello,

Is this single tcl command the only change you've made to the project? Have you modified the code in any way?

Does the project build when it is freshly checked out of the repository?

Also, according to p. 176 of the Vivado tcl [reference](https://www.xilinx.com/support/documentation/sw_manuals/xilinx2013_4/ug835-vivado-tcl-commands.pdf), it seems like it you may want to use **create_generated_clock** for putting constraints on the already defined physical clock in the project.

If you like, you can send us your project and I can try to debug the issue. If you'd like to keep the project private, then you can tell me which modifications you've made and I can try to reproduce the bug.

**Comment by Edgar-Conzen** [comment]
So I did some homework and got a little bit closer to the problem
I list it step by step:

>1. Freshly checked out the repository
>2. Created a Viviado 2017.1 project
>3. Added all source files from \trunk\rtl to project
>4. In Sources -> Hierarchy (tree view) selected 
>    rv : Orca(rtl) (orca.vhd) as Top Module (Set as Top)
>5. In orca.vhd changed generics 
>    AXI_ENABLE := 1
>    Family := "XILINX"
>6. Create a new constraint file clocks.xdc with one line only 
>    create_clock -name myclk -period 10 -waveform {0 5} [get_nets clk]
>    checked that this is correct (see link below)
>7. Synthesis works perfectly ok.
>8. Report Timing Summary works perfectly ok.

Now I changed ENABLE_EXT_INTERRUPTS  := 1 and the error is back

Background why I did not see this up to now:
In my previous attempts I selected orca_core.vhd as top.
 
So I set the generics in orca_core.vhd manually to the exact same values as in orca.vhd
At least I thought that I did this. 

But I missed the conditional in the instantiation (Line 454 in orca.vhd ) and set
NUM_EXT_INTERRUPTS := 1 (when it is in fact 0 by instatiation)
So I always tested with interrupts enabled and thus the critical warning was always there.

Sadly I cannot add the zipped project, because github rejects it. But essentially it's only your sources.

Thanks for the interest and support. I appreciate that you share your work!


Here is the above mentioned link which I found rather enlightening:
https://stackoverflow.com/questions/17456699/what-is-the-point-of-create-clock-command-in-fpga-design

Some rant about this topic that has nothing to do with you or your company ;-)

I think it's an odd and rather bad choice to name this constraint "create_clock".
At least it should be named "create_clock_object" or "specify_clock" or something alike.
Google shows A LOT of confusion about this topic and in my opinion this is founded in
the poor choice of the name. Especially to hardware developers "creating a clock" has
the very dedicated meaning of actually building hardware. In XDC it is instead used 
to **identify and characterise a signal as a clock to check the timing**. Just my opinion.

**Comment by rdeiaco** [comment]
I'm glad to hear that everything worked out.
Good luck with your project, and please let us know if you encounter any other issues.

**Comment by Edgar-Conzen** [comment]
Sorry for misleading you. My previous comment indeed sounded more like a success story.

Now I have zipped my project on my coputer at home and it seems to be uploading.
Would you please give it a try with Vivado 2017.1 ? 

I marked the few changes in orca.vhd with comments like " -- EC: ... " 

[orca_ec.zip](https://github.com/VectorBlox/orca/files/1106772/orca_ec.zip)

**Comment by rdeiaco** [comment]
Absolutely, thanks for the upload.

**Comment by rdeiaco** [comment]
Okay, I've taken a look at your project, and I can see the timing loop critical warnings. A few points about your project:

1) The Orca rtl that we have will not work in the way you have presently instantiated it. Right now you have the Orca as the top level in the directory, whereas in our project we have the Orca as part of a board design, in which it is hooked up to a clock from the Zynq processor system. I think this may be why you're running into combinational loops, as the 'clk' signal to the Orca is being driven by an I/O pin on the FPGA (based on the generated netlist). This would cause all of the Orca's synchronous logic to fail. So technically, the tools are right, there is a combinational loop, but it is because the 'clk' net of the Orca is not configured correctly.

2) I noticed that you're trying to run this project on Windows (based on the directory handles I'm seeing in the project). While the Orca project isn't specifically tested or supported on Windows, you should be able to get the project to work with minimal hassle. If you clone our repository and open Vivado in the systems/zedboard/ directory, you should be able to build the project by entering the following tcl commands in the tcl console:

`source project.tcl`
`init_project project/ project`
`create_bd project/ project design_1.tcl`

This should build everything that you need to get the Orca project up and running. You may need to modify the project further if you want the project to work on a different set of hardware then specified in the project.

To ensure there was no hdl issue, I updated the rtl files in our project with yours, and I was able to successfully build the project. The issue here is most likely the project-related issues outlined above.

Let me know if that helps, or if you need any additional assistance.

**Comment by Edgar-Conzen** [comment]
Thank you for your hints. A lot to test at the weekend!

Could you please leave this issue open for the next two weeks? I will definitely come back to you.

**Comment by rdeiaco** [comment]
No problem. I'll leave the issue open.
Good luck with debugging!

**Comment by Edgar-Conzen** [comment]
Back with a bit of delay. To my excuse, I've been to the FPGA-Kongress in Munich last week, where one of my personal highlights was Ted Marena from Microsemi, whose enthusiasm with RISC-V was very catching ;-)

Highly motivated I did the following so far:

- Build the project as you recommended
- Synthezise it -> without critical warning

- Reconfigure IP Orca_0 -> Enable External Interrupts (see attached pic)
- Tie new pin "global_interrupts[0:0]" to some other pin
  (arbitrarily chose Zynq's TTC0_WAVE0_OUT)
- Synthezise it again -> without critical warning

- Execute TCL-command "report_drc"

The critical error shows up:

LUTLP-1#1 Critical Warning
Combinatorial Loop Alert  
6 LUT cells form a combinatorial loop. This can create a race condition. Timing analysis may not be accurate. The preferred resolution is to modify the design to remove combinatorial logic loops. If the loop is known and understood, this DRC can be bypassed by acknowledging the condition and setting the following XDC constraint on any net in the loop: 'set_property ALLOW_COMBINATORIAL_LOOPS TRUE [net_nets <myHier/myNet>'. The cells in the loop are: design_1_i/Orca_0/U0/core/X/branch/instr_ARVALID_INST_0_i_3, design_1_i/Orca_0/U0/core/X/syscall/mepc[31]_i_7, design_1_i/Orca_0/U0/core/X/syscall/pc_corr_saved[31]_i_3, design_1_i/Orca_0/U0/core/X/suppress_valid_instr_out_i_2, design_1_i/Orca_0/U0/core/instr_fetch/two_cycle.valid_latch_i_2, design_1_i/Orca_0/U0/core/X/syscall/valid_input_latched_i_2.
Related violations: <none>

- Execute the following TCL-command
  show_schematic [get_cells -of [lindex [get_drc_violations] 0]]
- Probe the resulting schematic until the relevant cells and paths are shown

As Vivado states "If the loop is known and understood, this DRC can be bypassed..."

Are you probably aware of this loop? To fully understand it I would have to study the
whole implementation for maybe weeks.
[design_1_added_interrupt.pdf](https://github.com/VectorBlox/orca/files/1152273/design_1_added_interrupt.pdf)
[schematic_LUTs_only.pdf](https://github.com/VectorBlox/orca/files/1152274/schematic_LUTs_only.pdf)
[schematic_comb_loop.pdf](https://github.com/VectorBlox/orca/files/1152275/schematic_comb_loop.pdf)

**Comment by rdeiaco** [comment]
Hi Edgar,

I'll take a look and see if I can reproduce the error.

**Comment by rdeiaco** [comment]
Hi again,

To update you, you are right, there is a timing loop in the interrupt handler. I am testing a fix right now, and hope to push it to GitHub soon. Thanks for finding this bug!

**Comment by rdeiaco** [comment]
Hi Edgar,

I apologize for the delay.

I believe I've fixed the combinational loop in the interrupt controller, please try the sample project in the zedboard directory and let me know if you have any further issues.

**Comment by helaah** [comment]
Hi,
I m working on my graduation project but i m still newbie with vivado and FPGA designing and i had the same issue as M.Edgar, timing loop detected warning in a synthesis report, with a mixed Verilog/VHDL design. I want to know : 
- Timing loops may cause synthesis issue, hence wrong synthesizing result for my design ?
-  Whether i should modify my RTL code or just fix the problem with tcl commands?
I m so grateful if i get some help

 Thank you

**Comment by rdeiaco** [comment]
Hello,

If possible, could you please give us some more information on your project? Did you build our project as is or did you make any modifications? 
Did you build the project using our Makefile?
Where in the netlist are you seeing the timing loop?

Thanks,

Ryan

**Comment by Edgar-Conzen** [comment]
And... the critical warning about the timing loop is gone!

First I tried to synthesize the ZEDBOARD project but Vivado flagged two syntax errors
in edge_extender.vhd and utils.vhd. (There are several instances of these files in the checked
out directory tree and some of them contain only a single line like "../../utils.vhd")

So I simply created a new project, set the orca_core as top with interrupts enabled and added 
a create_clock constraint, because the timing loop was part of the orca_core hierarchy.
Synthesis went fine and even the TCL-command "report_drc" showed no critical warning.

Everything seems to work. Thank you very much for your support! I am looking forward 
to using your core.
